envoy/router/path_match_policy.h

```diff
@@ -0,0 +1,71 @@
+#pragma once
+
+#include "envoy/config/typed_config.h"
+
+#include "source/common/common/logger.h"
+
+#include "absl/status/statusor.h"
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace Router {
+
+/**
+ * Decides if the target route path is matching the provided pattern.
+ * Subclassing Logger::Loggable so that implementations can log details.
+ */
+class PathMatchPredicate : Logger::Loggable<Logger::Id::router> {
+public:
+  PathMatchPredicate() = default;
+  virtual ~PathMatchPredicate() = default;
+
+  /**
+   * Used to determine if the current url matches the predicate pattern.
+   *
+   * @param url current url of route
+   * @return valid if route url matches the predicate pattern.
+   */
+  virtual bool match(absl::string_view url) const PURE;
+
+  /**
+   * @return the match pattern of the predicate.
+   */
+  virtual std::string pattern() const PURE;
```

Comment:
It's problematic bc when downcastAndValidate takes the generic config and coverts its into the extension specific config it assumes strings are std::string. If its string_view it will still convert but pack the bits in wrong. Found this out in testing..

Reply:
I don't follow. Doesn't the implementation of this method live in PatternTemplateMatchPredicate and just returns its path_template_ member? I think the return type can just be a string_view and it should be automatically constructed from the std::string member on return.

The motivation here is to avoid copying strings, especially in the request path. string_views are designed to be cheap to create and don't require allocations. Another option is to have this return `const std::string&`, though I think our preference is string_view unless the caller would just need to repeatedly create actual std::strings from the view.

Reply:
This issue happens here:
`
  explicit PatternTemplateRewritePredicate(
      const envoy::extensions::path::rewrite::pattern_template::v3::PatternTemplateRewriteConfig&
          rewrite_config)
      : url_rewrite_pattern_(rewrite_config.path_template_rewrite()) {}
`

If I turn `rewrite_config.path_template_rewrite` into a string_view it creates a string_view but the string is scrambled and not created correctly

Reply:
The underlying storage (`url_rewrite_pattern_`) needs to be a std::string, but the `pattern()` method's return type can  be `string_view` (or `const std::string&`) to avoid copying the string each time pattern() is called. The choice between string_view and a std::string reference should be based on what the callers do with

Reply:
Got it. Changed over. Will do so in smaller PR as well.